tutorials/python/python-type-driven-development/tutorial.md

```diff
@@ -0,0 +1,207 @@
+
+To start using mypy, install it using pip globally or in your virtual environment.
+
+```bash
+pip install mypy
+```
+
+To perform static type checking to a source file:
+
+```bash
+mypy my_file.py
+```
+
+## Primitive types
+
+Python is a multi-paradigm dynamic language, so the type of a variable is determined based on the value.  Let's no get confused. Saying that Python is a dynamic language doesn't mean that has _weak typing_. Python is _strong typed_. Surprised? Check this out:
+
+```python
+movie = "Die Hard"
+movie = movie + 2
+# TypeError: cannot concatenate 'str' and 'int' objects
+```
+
+That said, Let’s begin this tutorial with this basic example:
+
+```python
+meat = "Ground beef"
+print(type(meat))
+# <class 'str'>
+
+weight_pounds = 0.5
+print(type(weight_pounds))
+# <class 'float'>
+```
+
+In the above example, we are using the `type` function to inspect the type representation of any variable. What if we include types to our first example code? Let’s check the following example:
+
+```python
+meat: str = "Ground beef"
+weight_pounds: str = 0.5
+```
+
+If we execute this file we will expect the same output as the previous one, but when we run `mypy` we will get the following message:
+
+```
+error: Incompatible types in assignment (expression has type "float", variable has type "str")
+```
+
+Did you catch the reason? The variable `weight_pounds` was defined as `str` but we were assigning a float number to it. We can start spotting bugs in every type check.
+
+I’m pretty sure you can understand this function:
+
+```python
+def make_hamburger(meat, number_of_meats):
+   return ["bread"] + [meat] * number_of_meats + ["bread"]
+
+
+print(make_hamburger("ground beef", 2))
+# ['bread', 'ground beef', 'ground beef', 'bread']
+```
+
+Let’s see how it goes with types:
+
+```python
+from typing import List
+
+
+def make_hamburger(meat: str, number_of_meats: int) -> List[str]:
+   return ["bread"] + [meat] * number_of_meats + ["bread"]
+```
+
+What happened here? We defined the function’s arguments types: string for `meat`, integer for `number_of_meats` and the return type which is a list of string values. Did you see the import at the beginning? Complex types must be imported from the `typing` library. We will cover some of them later.
+
+## Type Alias
+
+You can also define your custom type names for known structures, very useful to improve the readability. Let’s say if we want to consider a hamburger is a list of strings, we can define a type `Hamburger` in the following way:
+
+```python
+Hamburger = List[str]
+
+def make_hamburger(meat: str, number_of_meats: int) -> Hamburger:
+   return ["bread"] + [meat] * number_of_meats + ["bread"]
+```
+
+## Callables
+
+Are you doing functional programming? No problem, the typing library inclues a type for callables which accepts a two-dimensional list like `[[argument1_type, … argumentN_type], return_type]`.
```

Comment:
Is callable something that should be clear at this point from Python or would that require additional explanation?

Reply:
Good call. I'll add a brief description at the beginning.